DRAFT: Add mocking interfaces and implementations for testing

Ayiga opened the pull request:
<!-- Closes #<ISSUE_NUMBER> -->
<!-- These comments should help create a useful PR message, please delete any remaining comments before opening the PR. -->
<!-- If there is no issue number make sure to describe clearly *why* this PR is necessary. -->
<!-- Mention open questions, remaining TODOs, if any -->

### This PR:
<!-- Describe what this PR adds to this repo and why -->
<!-- E.g. -->
<!-- * Implements feature 1 -->
<!-- * Fixes bug 3 -->
Seeks to create a base for testing and verifying data flow from the `TransactionStreamer` to the Espresso Network.

This PR defines restricted interfaces to better define the actual methods being utilized in the process.  Additionally, this PR defines individual Mock implementations to simulate a simplified representation of the important systems that are working together.  These implementations allow for a simplified, controllable, testable framework for simulating these complex interactions.  Some of these Mock implementations also allow for simulation of network conditions / latency in some cases.

Additionally, this PR seeks to separate the ingrained Espresso additions to the `TransactionStreamer` into its own sub package within the repository.  The goal here is two-fold.  The first is to better provide clearer delineations between the original implementation and the additions that Espresso has made.  The second is to abstract away the process involved in communicating with Espresso, so different implementations / experiments can be attempted more easily.

<!-- ### This PR does not: -->
<!-- Describe what is out of scope for this PR, if applicable. Leave this section blank if it's not applicable -->
<!-- This section helps avoid the reviewer having to needlessly point out missing parts -->
<!-- * Implement feature 3 because that feature is blocked by Issue 4 -->
<!-- * Implement xyz because that is tracked in issue #123. -->
<!-- * Address xzy for which I opened issue #456 -->

<!-- ### Key places to review: -->
<!-- Describe key places for reviewers to pay close attention to -->
<!-- * file.rs, `add_integers` function -->
<!-- Or directly comment on those files/lines to make it easier for the reviewers -->

<!-- ### How to test this PR:  -->
<!-- Optional, uncomment the above line if this is relevant to your PR -->
<!-- If your PR is fully tested through CI there is no need to add this section -->
<!-- * E.g. `just test` -->

<!-- ### Things tested -->
<!-- Anything that was manually tested (that is not tested in CI). -->
<!-- E.g. building/running of docker containers. Changes to docker demo, ... -->
<!-- Especially mention anything untested, with reasoning and link an issue to resolve this. -->

<!-- Complete the following items before creating this PR -->
<!-- [ ] Issue linked or PR description mentions why this change is necessary. -->
<!-- [ ] PR description is clear enough for reviewers. -->
<!-- [ ] Documentation for changes (additions) has been updated (added).  -->
<!-- [ ] If this is a draft it is marked as "draft".  -->
<!-- [ ] For integration projects, make sure it won't break backward compatibility. -->

<!-- Details on maintaining backward compatibility for integration projects: -->
<!-- * Try to keep changes additive: add new, optional methods, flags, or parameters instead of modifying or removing existing functionality. -->
<!-- * If modification is necessary, it should be either a clear bug fix or guarded by a config/feature flag.  -->
<!-- * Follow Open Closed Principle and Interface Segregation Principle, clients should not be forced to depend on interfaces they do not use. -->

<!-- To make changes to this template edit https://github.com/EspressoSystems/.github/blob/main/PULL_REQUEST_TEMPLATE.md -->

Sneh1999 commented on arbnode/batch_poster.go:
```diff
@@ -584,37 +528,131 @@ func NewBatchPoster(ctx context.Context, opts *BatchPosterOpts) (*BatchPoster, e
-			return nil, err
-		}
-		verifier := espressotee.NewEspressoTEEVerifier(teeVerifier, opts.L1Reader.Client(), espressoTeeVerifierAddress)
+	// Espresso Config Round 2
```
Instead of making this a code block, we might as well make a separate function for adding all the espresso configs. What do you think?

Ayiga replied:
That's a very good idea.

I struggled with this code change.  The Espresso specific setup code was spread over two areas, and relied on initialization of more than one structure.  Yet it needs to affect the submission process in a meaningful way.  Ideally, this would all be setup at the time of the initialization of the `TransactionStreamer`, as the `TransactionStreamer` needs to potentially interact with the Espresso Submission process.

But, the Espresso Submission Process requires the `KeyManager`, which requires information from the `AccessList`, which requires the  the `DataPoster`.  It's a very weird chain of dependencies that are difficult to unwind.

Sneh1999 commented on arbnode/batch_poster.go:
```diff
@@ -751,25 +789,25 @@ func AccessList(opts *AccessListOpts) types.AccessList {
-	if b.streamer.EscapeHatchEnabled {
+
+	// We assume that submitter is not nil here.
+	submitter := b.streamer.espressoSubmitter
```
nit: rename to espressoSubmitter

Sneh1999 commented on arbnode/espresso.go:
```diff
@@ -0,0 +1,28 @@
+// It derives the method definitions from the execution package:
+// "github.com/offchainlabs/nitro/execution"
+//
+// It is defined separately here from the execution.ExecutionSequencer
```
This comment doesnt match the actual interface: https://github.com/OffchainLabs/nitro/blob/v3.6.7/execution/interface.go#L64-L73

Ayiga replied:
This is true.

This was **very recently** modified to be `ExecutionClient` instead of `ExecutionSequencer`.  So when I rebased, I missed that the i[nterface being utilized was changed.](https://github.com/EspressoSystems/nitro-espresso-integration/commit/e4bdf8121a35cf097c5c947a15246218b02a61af#diff-4c43892baa92d7cf1b422d5e9e7d5534ca1109dbecd5b647c8f8aaf57d3cc391R60)

Ayiga replied:
Actually, double checking this.  It seems like the `ExecutionClient` is a [requirement](https://github.com/OffchainLabs/nitro/blob/v3.6.7/execution/interface.go#L65) of `ExecutionSequencer`.  So this statement is still technically true.  But it would be better served to be more specific.

Sneh1999 commented on arbnode/espresso.go:
```diff
@@ -0,0 +1,28 @@
+package arbnode
```
nit: rename the file from espresso.go

Ayiga replied:
True.  This file originally contained all of the new interfaces I had defined for this refactor.  But Now that it only contains one, it can likely be added to the `transaction_streamer.go` file itself.

My aim is to try and minimize the amount of changes needed in `transaction_streamer.go`, to make our future changes / merges easier.  So keeping a separate file for this may still be desirable.

Sneh1999 commented on arbnode/espresso.go:
```diff
@@ -0,0 +1,28 @@
+	"github.com/offchainlabs/nitro/util/containers"
+)
+
+// TransactionStreamerExecutionSequencer defines the interface under which
```
I think this is an unnecessary optimization, I feel like its okay to add another method to the original interface. In future, we will have to manage both the interfaces. I dont see the benefit in that

Ayiga replied:
This is not really an optimization.  This is a restriction of the definition of the `github.com/offchainlabs/nitro/execution.ExecutionSequencer` interface to only those methods explicitly referenced / utilized within the `TransactionStreamer`.

This is done for convenience and nothing more. This change allows one to create a simplified mock of the `ExecutionSequencer` by only needing to implement these methods:

* `Reorg`
* `HeadMessageIndex`
* `MarkFeedStart`
* `ResultAtMessageIndex`
* `DigestMessage`
* `BlockNumberToMessageIndex`

Rather than the full set that the original interface provides:

* `DigestMessage`
* `Reorg`
* `HeadMessageIndex`
* `ResultAtMessageIndex`
* `MessageIndexToBlockNumber`
* `BlockNumberToMessageIndex`
* `SetFinalityData`
* `MarkFeedStart`
* `Maintenance`
* `Start`
* `StopAndWait`
* `Pause`
* `Activate`
* `ForwardTo`
* `SequenceDelayedMessage`
* `NextDelayedMessageNumber`
* `Synced`
* `FullSyncProgressMap`


The goal here is to keep the added code small rather than large with a lot of unnecessary code.  It's specifically targeting only the relevant sections of code, rather than the full interface.  This interface is more concerned with mocking implementations specifically for the `TransactionStreamer`, rather than the system as a whole.

zacshowa replied:
I do see how this helps tremendously on the mocking infra side. Is there a way to make it more apparent when the interface would drift between our new interface and the full execution client interface?

I could see a world in which "re-syncing the interfaces" becomes an task we have to deal with after doing code syncs before we can even run the tests. Maybe that's the appropriate time to perform a task like that though; I'm mainly thinking out loud.

Ayiga replied:
I think so.  We could add a type assertion check of one implementation to the other.  Something like this **should** work:

```go
var _ TransactionStreamerExecutionSequencer = execution.ExecutionSequencer(nil)
```

Ayiga replied:
I will say that explicitly modifying the `TransactionStreamer`'s `exec` field's type is less-than-ideal.  There may even be a middle ground here to introduce another wrapper that adheres to the full `execution.ExecutionClient` with some unimplemented errors for every method call.  Then add a unifying struct that defers to the reduced interface with overrides, but allows the passthrough call to the larger unimplemented case where it's not implemented explicitly.

Sneh1999 commented on arbnode/espresso/chain/delay_chain.go:
```diff
@@ -0,0 +1,58 @@
+
+// EspressoChainDelayed is a wrapper around an Espresso chain client that
+// introduces a delay before executing any method calls. This can be useful for
+// simulating network latency or for testing purposes.
```
nit: In arbitrum all the testing helper code goes inside the system_tests folder. I think we should put this file there as well

Sneh1999 commented on arbnode/espresso/chain/exponential_delay.go:
```diff
@@ -0,0 +1,63 @@
+package chain
```
same for this file, it should be put under system tests because this will only be used for testing

Ayiga replied:
We can move these utilities under the `system_tests` folder.  I would likely still keep it separated from the rest of the files under its own dedicated `espresso` folder to make it clear that is a purely additive change made by us to the original code base, and avoids namespace collisions within the package.

Sneh1999 commented on arbnode/espresso/chain/metrics.go:
```diff
@@ -0,0 +1,112 @@
+	"github.com/offchainlabs/nitro/arbnode/espresso"
+)
+
+type EspressoChainClientMetrics struct {
```
I think adding metrics is a good idea but it doesnt have to be a separate struct. We should think about if we can map the original espresso client in an interface and then if we want to enable metrics or not should be controlled by a boolean

Ayiga replied:
I disagree with that approach, and this suggestion.

Adding features like metrics, or logging to existing implementations by modifying their underlying code, can be destructive, as we are actually modifying the underlying code by definition.  This can lead to these underlying implementations being more difficult to modify without breaking some underlying concern / dependency.  This is especially valuable when modifying some other organization's or individual's code, but applies equally to modifying your own code.

In these cases, I favor [composition](https://en.wikipedia.org/wiki/Object_composition) and [encapsulation](https://en.wikipedia.org/wiki/Encapsulation_(computer_programming)) as an alternative to non-destructively add functionality.  This approach allows for the added mechanism to only have to be concerned with, and [responsible](https://en.wikipedia.org/wiki/Single-responsibility_principle) for, how to measure the results of the call, and not worry about how the actual implementation is done.  It provides clean boundaries for these [separations of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns).

Modifying the original structure to track these sort of metrics involves decisions such as picking where one makes the decision to initiate these measurements.  This is especially egregious / error prone when one allows users to opt-in to these features.  Doing so incurs a minimum of at least one conditional branch cost added per measuring point if performed naively.

If we focus on the composable approach instead, we have greater flexibility of opting in, and choosing where we measure, and add other such features.

Example:
Consider the following implementations of this Espresso Client with composition in the way I have performed it here:

* `C` -> The actual client implementation
* `D` -> A delay adding Middleware, that adds delays to the calls to fake latency due to network or other factors
* `M` -> A metrics tracking Middleware, that times the endpoints and stores each call for future reference.

With the composition approach, one can easily pick where the measurement occurs.  If I want to measure the metrics with the delay:

```
M(D(C))
```

If I want to measure just the underlying implementation:

```
D(M(C))
```

What if I don't want to collect metrics on the endpoint?

```
D(C)
```

Better yet, there is no conditional cost incurred by any opt-in portion, other than at the time of initialization.  Which makes the underlying code not subject to branch-prediction failure penalties at each call, but only at initialization time.


I'm more than happy to discuss this point further, if you'd like.

Sneh1999 commented on arbnode/espresso/chain/siphon_blocks.go:
```diff
@@ -0,0 +1,46 @@
+	ch    chan<- espresso_client.TransactionsInBlock
+}
+
+func NewSiphonBlocksWithTransactions(
```
what is a Siphon?

Sneh1999 commented on arbnode/espresso/submitter/original_submitter.go:
```diff
@@ -0,0 +1,779 @@
+package submitter
```
nit: can we change the name from `original_submitter` not sure what original means here?

Ayiga replied:
I named this as `original_submitter.go`, specifically as it contains the data structure named with `OriginalSubmitter`.

For the underlying reason of the name, please see my response on the comment of the struct's name itself.

Sneh1999 commented on arbnode/espresso/submitter/original_submitter.go:
```diff
@@ -0,0 +1,779 @@
+	espressoLastConfirmedPos     []byte = []byte("_espressoLastConfirmedPos") // contains the position of the last confirmed message
+)
+
+type EspressoOriginalSubmitter struct {
```
nit: same here, not sure what original means in this context? should prob rename

Ayiga replied:
This is named `EspressoOriginalSubmitter`, as I've made the effort to extract this submission to Espresso process into it's own struct.  It is named `OriginalSubmitter`, as it reflects the current implementation / strategy with no real logical modifications.

The purpose of this is to compare it with different strategies within the same test, so one might benchmark both.

Sneh1999 reviewed (commented):
Thanks for all the work here! 

Can we please split this PR into two parts:

1. That will add mock testing code
2. Which is doing major code changes to the core code

This PR is too big to review and changing a lot of code, so will appreciate if we do it in smaller PRs

Ayiga committed:
Rename local variables of `submitter` to `espressoSubmitter`

The feedback provided by @sneh1999 requests that the local variable `submitter` be
renamed to `espressoSubmitter` for clarity.  This request was made in this discussion
here:
https://github.com/EspressoSystems/nitro-espresso-integration/pull/661#discussion_r2201954711

ImJeremyHe committed:
Update golangci-lint in nix shell (#662)

Ayiga committed:
Relocate mocking code into `system_tests`

Based on feedback provided by @Sneh1999, all of the mocks for test related code have been relocated under the `system_tests` package.  These changes are primarily
motivated by these requests / discussions:

https://github.com/EspressoSystems/nitro-espresso-integration/pull/661#discussion_r2201960862
https://github.com/EspressoSystems/nitro-espresso-integration/pull/661#discussion_r2201962054
https://github.com/EspressoSystems/nitro-espresso-integration/pull/661#discussion_r2201959621

Due to the moved files, imports had to be updated and moved around.

Ayiga committed:
Refactor test setup to allow for easier mock setup

The Test environment that is being setup to measure performance is quite flexible
but it is not obvious how to setup effectively.  It currently requires many individual
calls and configurations that allow for greater flexibility, but not the ease of use and
setup.  Ideally, we would like the ability to setup a mock environment with some
default configuration without the need to specify things.

In other words, the ideal environment should be able to be setup with the invocation of
a single function, while still allowing for specific configurations to be modified.

This change refactors the aspects of this setup in an effort to provide the caller with
maximum flexibility, and control.

Ayiga commented:
Closing as it has been replaced by #668
